tests/utils/utils.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/ethereum/go-ethereum/log"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	DefaultTeleporterTransactionGas       uint64 = 200_000
+	DefaultTeleporterTransactionGasFeeCap        = big.NewInt(225 * params.GWei)
+	DefaultTeleporterTransactionGasTipCap        = big.NewInt(params.GWei)
+	DefaultTeleporterTransactionValue            = common.Big0
+)
+
+type SendCrossChainMessageEvent struct {
+	DestinationChainID ids.ID
+	MessageID          *big.Int
+	Message            teleporter.TeleporterMessage
+}
+
+type ReceiveCrossChainMessageEvent struct {
+	OriginChainID ids.ID
+	MessageID     *big.Int
+	Message       teleporter.TeleporterMessage
+}
+
+// Teleporter contract sendCrossChainMessage input type
+type SendCrossChainMessageInput struct {
+	DestinationChainID      ids.ID
+	DestinationAddress      common.Address
+	FeeInfo                 FeeInfo
+	RequiredGasLimit        *big.Int
+	Message                 []byte
+	AllowedRelayerAddresses []common.Address
+}
+
+type FeeInfo struct {
+	ContractAddress common.Address
+	Amount          *big.Int
+}
+
+//
+// Test utility functions
+//
+
+// Subscribes to new heads, sends a tx, and waits for the new head to appear
+// Returns the new head
+func SendTransactionAndWaitForAcceptance(
+	ctx context.Context,
+	wsClient ethclient.Client,
+	tx *types.Transaction) *types.Receipt {
+
+	newHeads := make(chan *types.Header, 1)
+	subA, err := wsClient.SubscribeNewHead(ctx, newHeads)
+	Expect(err).Should(BeNil())
+	defer subA.Unsubscribe()
+
+	err = wsClient.SendTransaction(ctx, tx)
+	Expect(err).Should(BeNil())
+	// Wait for the transaction to be accepted
+	<-newHeads
+
+	receipt, err := wsClient.TransactionReceipt(ctx, tx.Hash())
+	Expect(err).Should(BeNil())
+	Expect(receipt.Status).Should(Equal(types.ReceiptStatusSuccessful))
+
+	return receipt
+}
+
+func HttpToWebsocketURI(uri string, blockchainID string) string {
+	return fmt.Sprintf("ws://%s/ext/bc/%s/ws", strings.TrimPrefix(uri, "http://"), blockchainID)
+}
+
+func HttpToRPCURI(uri string, blockchainID string) string {
+	return fmt.Sprintf("http://%s/ext/bc/%s/rpc", strings.TrimPrefix(uri, "http://"), blockchainID)
+}
+
+func GetURIHostAndPort(uri string) (string, uint32, error) {
+	// At a minimum uri should have http:// of 7 characters
+	Expect(len(uri)).Should(BeNumerically(">", 7))
+	if uri[:7] == "http://" {
+		uri = uri[7:]
+	} else if uri[:8] == "https://" {
+		uri = uri[8:]
+	} else {
+		return "", 0, fmt.Errorf("invalid uri: %s", uri)
+	}
+
+	// Split the uri into host and port
+	hostAndPort := strings.Split(uri, ":")
+	Expect(len(hostAndPort)).Should(Equal(2))
+
+	// Parse the port
+	port, err := strconv.ParseUint(hostAndPort[1], 10, 32)
+	if err != nil {
+		return "", 0, fmt.Errorf("failed to parse port: %w", err)
+	}
+
+	return hostAndPort[0], uint32(port), nil
+}
+
+func CreateTeleporterTransaction(
+	ctx context.Context,
+	source SubnetTestInfo,
+	destination SubnetTestInfo,
```

Comment:
Instead of taking the `destination SubnetTestInfo` as a parameter here, I think it would be more useful to have it be a `SendCrossChainMessageInput`. That way, this utility can be used generally to creating transactions that send arbitrary messages.

Reply:
Changed 👍